waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+                                                    :saml-auth-data saml-auth-data})
+     :status 200}))
+
+(defn saml-auth-redirect-handler
+  "Endpoint for POSTs to Waiter with IdP-signed credentials. If signature is valid, return principal and original request."
+  [request {:keys [password]}]
+  {:pre [(not-empty password)]}
+  (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+    (let [{:keys [not-on-or-after redirect-url saml-principal]}
+          (try
+            (utils/base-64-string->map saml-auth-data password)
+            (catch Exception e
+              (throw (ex-info "Could not parse saml-auth-data." {:status 400
+                                                                 :saml-auth-data saml-auth-data
+                                                                 :inner-exception e} e))))
+          t-now (t/now)
+          _ (when-not (t/before? t-now not-on-or-after)
+              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                              {:status 400
+                               :saml-assertion-not-on-or-after not-on-or-after
+                               :t-now t-now})))
+          auth-cookie-expiry-date (t/min-date not-on-or-after (t/plus t-now (t/days 1)))
+          auth-cookie-age-in-seconds (-> t-now
+                           (t/interval auth-cookie-expiry-date)
+                           (t/in-seconds))
+          {:keys [authorization/principal authorization/user] :as auth-params-map}
+          (auth/auth-params-map saml-principal)]
+      (auth/handle-request-auth (constantly {:status 303
+                                             :headers {"location" redirect-url}
+                                             :body ""})
+                                request principal auth-params-map password auth-cookie-age-in-seconds))
+    (throw (ex-info "Missing saml-auth-data from SAML authenticated redirect message"
+                    {:status 400}))))
+
+(defn create-request
+  "Return XML elements that represent a SAML 2.0 auth request."
+  [time-issued saml-service-name saml-id acs-url idp-uri]
+  (str
+    (hiccup.page/xml-declaration "UTF-8")
+    (hiccup/html
+      [:samlp:AuthnRequest
+       {:xmlns:samlp "urn:oasis:names:tc:SAML:2.0:protocol"
+        :ID saml-id
+        :Version "2.0"
+        :IssueInstant time-issued
+        :ProtocolBinding "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+        :ProviderName saml-service-name
+        :IsPassive false
+        :Destination idp-uri
+        :AssertionConsumerServiceURL acs-url}
+       [:saml:Issuer
+        {:xmlns:saml "urn:oasis:names:tc:SAML:2.0:assertion"}
+        saml-service-name]])))
+
+(def instant-format (f/formatters :date-time-no-ms))
+
+(defn make-issue-instant
+  "Converts a date-time to a SAML 2.0 time string."
+  []
+  (f/unparse instant-format (t/now)))
+
+(let [opensaml-bootstrapped-atom (atom false)]
```

Comment:
Why do we need this atom?

Reply:
want to call org.opensaml.DefaultBootstrap/bootstrap only once

Reply:
`create-request-factory!` is called from the factory function which is called only once?

Reply:
maybe was thinking for unit tests. but yeah, called once